### ISSUE TITLE
[cc] MOCK_METHOD() type macros can break syntax outside of the macro call

### ISSUE THREAD

**Issue by mtaran-google** [codemirror/google-modes#128]
```cpp

class MockFoo : public ns::ThreadSafeFoo {
 public:
  MockFoo() : ns::ThreadSafeFoo("") {}

  MOCK_METHOD3(DoFoo, bool(const ns::Something&,
                           const std::vector<int>&, const string));
  MOCK_METHOD3(DoSomethingElse, bool(const ns::Something&,
                                     const std::vector<int>&, const string));

 private:
  DISALLOW_COPY_AND_ASSIGN(MockFoo);
};

class MockBar : public Bar {
 public:
  MockBar()
      : Bar(nullptr, "", "", absl::Seconds(1)) {}
  ~MockBar() override {}

  MOCK_METHOD2(DoBar, void(absl::string_view, absl::string_view));

 private:
  DISALLOW_COPY_AND_ASSIGN(MockBar);
};
```

note the `const string` in the DoFoo mock. If you change it to `const string&`, highlighting in the MockBar definition below changes.

**Comment by marijnh** [comment]
The mode currently parses macro arguments as expressions, which I guess isn't such a good idea. Would it be okay to switch to a simple-minded parsing of balanced brackets inside of macro argument lists, where it would recognize keywords, strings, numbers, and built-in types, but nothing else? That would mean that highlighting and indentation of constructs inside such an argument list would suffer, since the mode doesn't know what they are (it can't, due to the grammatical wormhole created by the macro).

**Comment by mtaran-google** [comment]
That should be worth a try. Would it be possible to also highlight usages of locals?

Also, if keywords are going to be highlighted in this context (even when they are in e.g. type position), maybe it makes sense to highlight keywords instead of/together with types in general? I believe this is how Eclipse, VS, and CLion also do it (for better or worse).

**Comment by marijnh** [comment]
> Would it be possible to also highlight usages of locals?

We could treat all identifiers as variables in this context, but that'd also highlight type names as variables, which might be annoying.

> maybe it makes sense to highlight keywords instead of/together with types in general?

I'm not sure what you mean. Can you give an example?

**Comment by marijnh** [comment]
> I'm not sure what you mean here

I'm saying that type names inside of macro argument lists will be treated as variable names. Is that acceptable?

> So instead, the CM mode could decorate int as both keyword and type. Does that make sense?

It could, but highlighting types uniformly seems like a good thing.

**Comment by mtaran-google** [comment]
> I'm saying that type names inside of macro argument lists will be treated as variable names. Is that acceptable?

If by that you mean "they will get no highlighting, unless they are already local variables, in which case they'll get the local-variable-usage highlighting", then yes, that should be good.

> It could, but highlighting types uniformly seems like a good thing.

Yes, it also seemed like a good thing to me too. However, since it's not possible to correctly highlight type usages inside of macros, it's not possible to highlight type usages uniformly in all contexts. Given that this is the case, it would be *more* consistent to 
A) always highlight keywords (this should apply when the mode is recovering from something unparseable too, btw), and only after that, it can
B) try to highlight types when it is able to do so

TL;DR: Users will be able to consistently recognize keywords, and also will still be able to recognize types when the grammar mode supports it.

Syntax | Now | With change
------------ | ------- | ------
vector<**int**> a; | `type` | `type keyword`
**vector**\<int> a; | `type` | `type`
MACRO(vector<**int**>) | `` | `keyword`
MACRO(**vector**\<int>) | `` | ``

**Comment by marijnh** [comment]
Attached patch changes the way macro arguments are parsed.

Regarding the type keywords, that turned out to be non-trivial to change, so I'm going to consider that a new feature and/or something that should have been specified in the requirements.
